[WIP] Add Storage Transformers to Core Spec & Sharding v1.0

jstriebel opened the pull request:
To get an auto-reloading server, run (using venv)
```
python3 -m venv .venv
. .venv/bin/activate
pip install -r docs/requirements
pip install sphinx-autobuild
sphinx-autobuild docs docs/_build/html
```

Adapting the abstract store interface is still TODO, it needs support for partial and bundled access for multiple chunks.

Some open questions remain, e.g.
* Should we introduce the abstract concept of storage transformers? I think this would be helpful for many other use cases.
* Storage transformers vs extensions:
  * Should storage transformers have a `must_understand` flag?
  * Can storage transformers be extensions?
  * Are all of them extensions?
  * Should sharding be defined as an extension?
  
Those could be up for discussion in the final PR.

jstriebel committed:
WIP

jstriebel committed:
finished first draft of sharding spec

jstriebel committed:
minor fixes

valentin-pinkau commented on docs/storage_transformers/sharding/v1.0.rst:
```diff
@@ -0,0 +1,185 @@
+For all entries that are part of the same shard the key is changed to the
+shard-key and the values are combined in the `Binary shard format`_ described
+below. The new shard-key is the chunk key divided by ``chunks_per_shard`` and
+floored per dimension. E.g. for ``chunks_per_shard=[32, 2]``, the chunk grid
```
Maybe start the example similar as above with: For example...

valentin-pinkau commented on docs/storage_transformers/sharding/v1.0.rst:
```diff
@@ -0,0 +1,185 @@
+    ``[32, 2]``, ``[32, 4]``, ``[64, 2]`` or ``[96, 18]``.
+
+
+Key & value transformation
```
This is basically the storage transformer implementation for sharding, right?
Maybe reflect that in this heading.

valentin-pinkau replied:
Key & value transformation with sharding storage transformer?

jstriebel replied:
Instead I added `Storage transformer implementation` as another heading and put the other two as subsections:

```
Storage transformer implementation
==================================

Key & value transformation
--------------------------

…


Binary shard format
-------------------

…
```

valentin-pinkau commented on docs/storage_transformers/sharding/v1.0.rst:
```diff
@@ -0,0 +1,185 @@
+* **Append-only**: Any chunk to write is appended to the existing shard, followed by an updated index.
+
+Any configuration parameters for the write strategy must not be part of the metadata document,
+in a shard I'd propose to use Morton order, but this can easily be changed and customized, since any order can be read.
+
```
This sounds like an opinion, which should not be part of the spec, what do you think?

jstriebel replied:
Good point, that was just a left-over from a copy and paste, removed it and adapted the previous sentence

valentin-pinkau commented on docs/storage_transformers/sharding/v1.0.rst:
```diff
@@ -0,0 +1,185 @@
+
+In the indexed binary format chunks are written successively in a shard, where
+unused space between them is allowed, followed by an index referencing them.
+The index holds an `offset, length` pair of little-endian uint64 per chunk,
```
At which offset do I find the index?

jstriebel replied:
At the end of the file, I added an explanation.

valentin-pinkau reviewed (commented):
Very nice! I left some comments :)

jstriebel committed:
apply feedback

jstriebel committed:
adapt abstract store interface

jstriebel commented:
@valentin-pinkau Danke dir für's Feedback! Kannst du vielleicht auch nochmal über den letzten Commit 28a9023 schauen, ob das für dich so Sinn ergibt?

normanrz commented on docs/storage_transformers/sharding/v1.0.rst:
```diff
@@ -0,0 +1,191 @@
+Motivation
+==========
+
+Sharding decouples the concept of chunks from storage keys, which become shards.
```
That is not entirely true. Shards hold 0 or more chunks, but not partial chunks.

A definition of sharding could be "co-locating multiple chunks within a file/storage object"

normanrz commented on docs/storage_transformers/sharding/v1.0.rst:
```diff
@@ -0,0 +1,191 @@
+  chunks, whereas
+- storage often is optimized for larger data per entry and fewer entries, e.g.
+  as restricted by the file block size and maximum inode number for typical
+  file systems.
```
```suggestion
- it becomes inefficient or impractical to store a large number of chunks in single files or objects due to the design constraints of the underlying storage, e.g. as restricted by the file block size and maximum inode number for typical file systems.
```

normanrz commented on docs/storage_transformers/sharding/v1.0.rst:
```diff
@@ -0,0 +1,191 @@
+Sharding decouples the concept of chunks from storage keys, which become shards.
+This is helpful when the requirements for those don't align:
+
+- Compressible units of chunks often need to be read and written in smaller
+  chunks, whereas
```
```suggestion
- Chunk sizes need to be small for read efficiency requirements, e.g. for data streaming in browser-based visualization software, whereas
```

writing is not really facilitated through sharding.

normanrz commented on docs/storage_transformers/sharding/v1.0.rst:
```diff
@@ -0,0 +1,191 @@
+  as restricted by the file block size and maximum inode number for typical
+  file systems.
+
+This does not necessarily fit the access patterns of the data, so chunks might
+need to be smaller than one storage key. In those cases sharding decouples those
+entities. One shard corresponds to one storage key, but can contain multiple chunks:
```
```suggestion
This does not necessarily fit the access patterns of the data, so chunks might
need to be smaller than the minimum size of one storage key. In those cases sharding decouples those
entities. One shard corresponds to one storage key, but can contain multiple chunks:
```

normanrz commented on docs/storage_transformers/sharding/v1.0.rst:
```diff
@@ -0,0 +1,191 @@
+The storage transformer protocol defines the abstract interface to be the same
+as the :ref:`abstract-store-interface`.
+
+The Zarr store interface is defined in terms of `keys` and `values`,
```
```suggestion
The Zarr store interface is defined as a mapping of `keys` and `values`,
```

normanrz commented on docs/storage_transformers/sharding/v1.0.rst:
```diff
@@ -0,0 +1,191 @@
+of bytes. A key-value pair is called `entry` in the following part.
+
+This sharding transformer only adapts entries where the key starts
+with `data/root`, as they indicate data keys for array chunks. All other
```
where does that come from? is that a constant or an example? can this be referenced to another part of the spec?

jstriebel replied:
That's part of the storage-key spec, a constant. I added a reference

normanrz commented on docs/storage_transformers/sharding/v1.0.rst:
```diff
@@ -0,0 +1,191 @@
+
+In the indexed binary format chunks are written successively in a shard, where
+unused space between them is allowed, followed by an index referencing them.
+The index is placed at the end of the file and has a length of 16 bytes per chunk
```
```suggestion
The index is placed at the end of the file and has a length of 16 bytes multiplied by the number of chunks
```

normanrz commented on docs/storage_transformers/sharding/v1.0.rst:
```diff
@@ -0,0 +1,191 @@
+
+Any configuration parameters for the write strategy must not be part of the metadata document,
+they need to be configured at runtime, as this is implementation specific.
+
```
I think what is missing is a read and write protocol, i.e. how does reading/writing work in an abstract algorithm.

normanrz reviewed (commented):
Most of it reads great!
- I think a read/write protocol is missing from the sharing spec.
- The "Motivation" doesn't sound convincing/concise enough to me. I added some suggestions, not sure if they help 😅

jstriebel committed:
Apply suggestions from code review

Co-authored-by: Norman Rzepka <code@normanrz.com>

jstriebel committed:
rename multi methods to partial_values

jstriebel committed:
improve motivation, add reference

jstriebel committed:
add api implementation protocol

jstriebel commented:
Thanks for the great feedback!
> I think a read/write protocol is missing from the sharing spec.

I added an `API implementation` chapter now covering this.

> The "Motivation" doesn't sound convincing/concise enough to me. I added some suggestions, not sure if they help sweat_smile

Yep, I rewrote it quite a bit again, I personally think it's better now. I'm not even sure if it should be part of the spec at all, but I think it's still helpful. What do you think?

Could you please give those two parts "Motivation" & "API implementation" in `docs/storage_transformers/sharding/v1.0.rst` another look @normanrz & @valentin-pinkau? Thanks a ton :pray:

normanrz commented on docs/storage_transformers/sharding/v1.0.rst:
```diff
@@ -0,0 +1,279 @@
+read efficiency requirements, for example to stream data in browser-based visualization software.
+
+Therefore, chunks may need to be smaller than the minimum size of one storage key.
+In those cases it is required to store objects at a more coarse granularity than reading chunks.
```
```suggestion
In those cases it is efficient to store objects at a more coarse granularity than reading chunks.
```

normanrz reviewed (approved):
I think this is good for sending out for public feedback!

> I'm not even sure if it should be part of the spec at all, but I think it's still helpful. What do you think?

I think this should definitely be in the spec in order to convince people that sharding is important to implement.

The `API implementation` could even be a bit more high-level for my taste, but let's see what the feedback is.

jstriebel committed:
Update docs/storage_transformers/sharding/v1.0.rst

Co-authored-by: Norman Rzepka <code@normanrz.com>

valentin-pinkau commented on docs/storage_transformers/sharding/v1.0.rst:
```diff
@@ -0,0 +1,279 @@
+  referenced. In this case a shard can be constructed according to the
+  `Binary shard format`_ directly.
+  For all other keys, request the indices from the underlying store using
+  ``get_partial_values``. All chunks that are not updated completely and
```
how does the `get_partial_values` function return the indices?

jstriebel replied:
The index format is described above.

valentin-pinkau replied:
So you use (store or next transformer).get_partial_values to read the index to perform the operation of the transformer.get_partial_values function?
To me, this is confusing as they are called the same but are defined on the transformer and on the store, right?

valentin-pinkau replied:
But I guess it makes sense :)

jstriebel replied:
> So you use (store or next transformer).get_partial_values to read the index to perform the operation of the transformer.get_partial_values function?

Exactly :+1: 

> To me, this is confusing as they are called the same but are defined on the transformer and on the store, right?

Yes, a bit, but that's also part of the beauty, as nobody cares if the "next" store is actually the store or just the next transformer. I tried to always use the term `underlying store` refering to the next transformer or store, as described in the paragraph above in the spec:
> The term `underlying store` references either the next storage transformer
in the stack or the actual store if this transformer is the last one in the

However, it's still confusing, but I have no idea how to improve it. Do you have a suggestion?

valentin-pinkau reviewed (commented):
Very nice, I have one question:
